klippy/extras/bed_mesh.py

```diff
@@ -243,31 +283,80 @@ def finalize(self, offsets, positions):
+        if len(positions) % expected_point_count:
             raise self.gcode.error(
                 ("bed_mesh: Invalid probe table length:\n"
                  "Sampled table length: %d") % len(positions))
-        samples = len(positions) / z_table_len
-        # Populate the organized probed table
-        for i in range(z_table_len):
-            y_position = i / x_cnt
-            x_position = 0
-            if y_position & 1 == 0:
-                # Even y count, x probed in positive directon
-                x_position = i % x_cnt
-            else:
-                # Odd y count, x probed in the negative directon
-                x_position = (x_cnt - 1) - (i % x_cnt)
-            idx = i * samples
-            end = idx + samples
-            avg_z = sum(p[2] for p in positions[idx:end]) / samples
-            self.probed_z_table[y_position][x_position] = \
-                avg_z - z_offset
+        samples = len(positions) / expected_point_count
+
+        if self._bed_shape == 'square':
+            # Populate the organized probed table
+            for i in range(z_table_len):
+                y_position = i / x_cnt
+                x_position = 0
+                if y_position & 1 == 0:
+                    # Even y count, x probed in positive directon
+                    x_position = i % x_cnt
+                else:
+                    # Odd y count, x probed in the negative directon
+                    x_position = (x_cnt - 1) - (i % x_cnt)
+                idx = i * samples
+                end = idx + samples
+                avg_z = sum(p[2] for p in positions[idx:end]) / samples
+                self.probed_z_table[y_position][x_position] = \
+                    avg_z - z_offset
+
+        elif self._bed_shape == 'circle':
+            #Reconcile the probed points and put them into the expected z_table
+            #To do this we need to essentually generate the points again
+            # since the points returned from the probing need to be placed
+            # in their correct spots (we need to match things up)
+
+            position_available_count = 0
+            position_total_count = 0
+
+            for i in range(y_cnt):
+                pos_y = -self._radius + self._distance_y * i
+                for j in range(x_cnt):
+                    pos_x = -self._radius + self._distance_x * j
+
+                    distance_from_center = math.floor(math.sqrt(pos_x * pos_x
+                                                        + pos_y * pos_y))
+
+                    y_position = position_total_count / x_cnt
+                    x_position = 0
+                    if y_position & 1 == 0:
+                        # Even y count, x probed in positive directon
+                        x_position = position_total_count % x_cnt
+                    else:
+                        # Odd y count, x probed in the negative directon
+                        x_position = (x_cnt - 1) - (position_total_count % x_cnt)
+
```

Comment:
Here is where I get confused.  Points are currently generated in a linear fashion, negative to positive.   While I understand your reasoning for "regenerating" the points, they are being added to the probed_z_table as if they were probed in a zigzag fashion.

Regardless I suggest that we simply sort the incoming array of probed values by their x and y coordinates.  Its much easier to do this now that bed_mesh no longer as to worry about averaging multiple samples, that is now handled by ProbePointsHelper.  If we do it that way, we do not need separate implementations for circular and square beds, and there is no need to regenerate points.   I'll work up something and post a link in this thread when I get a chance.